cmd/clusterctl/client/cluster/crd_migration.go

```diff
@@ -0,0 +1,224 @@
+	}
+
+	currentStatusStoredVersions := sets.NewString(currentCRD.Status.StoredVersions...)
+
+	// If the new CRD still contains all current stored versions, nothing to do
+	// as no previous storage version will be dropped.
+	if newVersions.HasAll(currentStatusStoredVersions.List()...) {
+		log.V(2).Info("CRD migration check passed", "name", newCRD.Name)
+		return false, nil
+	}
+
+	// Otherwise a version that has been used as storage version will be dropped, so it is necessary to migrate all the
+	// objects and drop the storage version from the current CRD status before installing the new CRD.
+	// Rif https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#writing-reading-and-updating-versioned-customresourcedefinition-objects
+	// Note: We are simply migrating all CR objects independent of the version in which they are actually stored in etcd.
+	// This way we can make sure that all CR objects are now stored in the current storage version.
+	// Alternatively, we would have to figure out which objects are stored in which version but this information is not
+	// exposed by the apiserver.
+	storedVersionsToDelete := currentStatusStoredVersions.Difference(newVersions)
+	storedVersionsToPreserve := currentStatusStoredVersions.Intersection(newVersions)
+	log.Info("CR migration required", "kind", newCRD.Spec.Names.Kind, "storedVersionsToDelete", strings.Join(storedVersionsToDelete.List(), ","), "storedVersionsToPreserve", strings.Join(storedVersionsToPreserve.List(), ","))
+
+	if err := m.migrateResourcesForCRD(ctx, currentCRD, currentStorageVersion); err != nil {
+		return false, err
+	}
+
+	if err := m.patchCRDStoredVersions(ctx, currentCRD, storedVersionsToPreserve.List()); err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (m *crdMigrator) migrateResourcesForCRD(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition, currentStorageVersion string) error {
+	log := logf.Log
+	log.Info("Migrating CRs, this operation may take a while...", "kind", crd.Spec.Names.Kind)
+
+	list := &unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: currentStorageVersion,
+		Kind:    crd.Spec.Names.ListKind,
+	})
+
+	var i int
+	for {
+		if err := retryWithExponentialBackoff(newReadBackoff(), func() error {
+			return m.Client.List(ctx, list, client.Continue(list.GetContinue()))
+		}); err != nil {
+			return errors.Wrapf(err, "failed to list %q", list.GetKind())
+		}
+
+		for i := range list.Items {
+			obj := list.Items[i]
+
+			log.V(5).Info("Migrating", logf.UnstructuredToValues(obj)...)
+			if err := retryWithExponentialBackoff(newWriteBackoff(), func() error {
+				return handleMigrateErr(m.Client.Update(ctx, &obj))
```

Comment:
Note to myself: verify how/if that changes managed fields of topology owned resources

Reply:
Looks like no changes to managed fields